libtock/crt0.c

```diff
@@ -78,22 +78,25 @@ void _start(void* app_start __attribute__((unused)),
     "movs r5, #7\n"
     "bic  r4, r4, r5\n"         // r4 = (mem_start + myhdr->stack_size + 7) & ~0x7
     //
-    // Compute the heap size
+    // Compute the app data size and where initial app brk should go.
+    // This includes the GOT, data, and BSS sections. However, we can't be sure
+    // the linker puts them back-to-back, but we do assume that BSS is last
+    // (i.e. myhdr->got_start < myhdr->bss_start && myhdr->data_start <
+    // myhdr->bss_start). With all of that true, then the size is equivalent
+    // to the end of the BSS section.
     //
-    // uint32_t heap_size = myhdr->got_size + myhdr->data_size + myhdr->bss_size;
-    "ldr  r5, [r0, #8]\n"       // r5 = myhdr->got_size
-    "ldr  r6, [r0, #20]\n"      // r6 = myhdr->data_size
-    "ldr  r7, [r0, #28]\n"      // r6 = myhdr->bss_size
-    "add  r5, r5, r6\n"         // r5 = got_size + data_size
-    "add  r5, r5, r7\n"         // r5 = got_size + data_size + bss_size
+    // uint32_t appdata_size = myhdr->bss_start + myhdr->bss_size;
+    "ldr  r5, [r0, #24]\n"      // r6 = myhdr->bss_start
+    "ldr  r6, [r0, #28]\n"      // r6 = myhdr->bss_size
+    "add  r5, r5, r6\n"         // r5 = bss_start + bss_size
```

Comment:
As discussed (though not agreed upon) on Slack, I think we need to substract `got_start` here.

Both `bss_start` and `got_start` are defined at link time to be location in memories of the BSS and GOT sections, respectively, _relative to the memory start used for linking_.

Currently, our linker script uses `0x0` as the beginning of RAM, so, in practice, those addresses _will_ be an offset from the start of RAM, regardless of where it starts. However, I don't think this is guaranteed. If `got_start` were guaranteed to be 0, we don't need it in the header in the first place.

For example, the linker _could_ place the GOT at an offset from `0x0`, which would push _everything_ forward by the offset. In our loader (this file), we _do not_ take that into account. We simply take the GOT section and stick it immediately above the stack.

To see this, try sticking a garbage section _in RAM_ right before the got section in the linker script. Alternatively, change the `RAM_START` in the linker script to a non-zero value. If you do that, this will break.